src/main/cpp/src/histogram.cu

```diff
@@ -97,7 +97,13 @@ struct fill_percentile_fn {
       return;
     }
 
-    output[idx] = (higher - position) * lower_element + (position - lower) * higher_element;
+    // Using `volatile` qualifier to prevent the compiler from combining `lower_part` and
+    // `upper_part` which may lead to output with different round-off error.
+    volatile double const lower_part =
+        (static_cast<double>(higher) - position) * static_cast<double>(lower_element);
+    volatile double const higher_part =
+        (position - static_cast<double>(lower)) * static_cast<double>(higher_element);
+    output[idx] = lower_part + higher_part;
```

Comment:
:+1: It is interesting to know that this sort of thing is still happening.

There are still `TODO`s in the original `libcudf` device code that allude to a similar problem, purported to be eventually resolved in `CUDA 10.2`. :]